data/network/src/commonMain/kotlin/social/androiddev/common/network/MastodonApiImpl.kt

```diff
@@ -35,4 +38,25 @@ class MastodonApiImpl(
             Result.failure(exception = exception)
         }
     }
+
+    override suspend fun createApplication(
+        clientName: String,
+        redirectUris: String,
+        scopes: String,
+        website: String?
+    ): Result<Application> {
```

Comment:
Is this the Kotlin result type? is it out of experimental internal apis now?

Reply:
Yes it is - I think for now we have this as the wrapping return type but this may possibly change. I actually thought you couldn't use it as a return type but I guess that changed?

Reply:
It was internal for a while but I guess they made it public. 

The unfortunate thing about Result is that your failure has to extend Throwable. So we can either keep result or write an `Either<L, R>` type which is quite simple thanks to `Nothing` and extension functions.

Reply:
Personally I find _Either_ too abstract and I prefer _Result_, as it is clear about what it is and what it contains in each case/side. However, as I don't like using Throwables for the error side (I generally use enums and sealed classes), I have a custom result class (it can be further extended).  
What do you think? https://gist.github.com/lotdrops/cc0714ffc145f46d97539bf4b03fa050

Reply:
I was going to suggest something similar - as I don't think we should leak lower level errors/exceptions. These can be mapped to a type that the domain/app can consume and determine what to do if more granular handling is needed. 

Here is a gist of `APIResponse` I created, a return type for all API calls that maps the response code, error and body to a specific type.

https://gist.github.com/crocsandcoffee/4d88fde78bd61aeabbca4ab900deb210

Reply:
In my projects I use MyResult in all layers, and with the map utils I map errors and success from data to domain models, and from domain to UI models (when I have different UI models).   
For API, I create a MyResult.Error with `enum class GenericApiError { NoInternet, Network, Server, Unknown }`.
I also have another for specific errors:```

sealed class CompleteApiError<in T>
object NoInternetApiError : CompleteApiError<Any?>()
object NetworkApiError : CompleteApiError<Any?>()
class ServerApiError(
    override val message: String = "",
) : CompleteApiError<Any?>(), MessageError
object UnknownApiError : CompleteApiError<Any?>()
class ExpectedApiError<T>(
    val which: T,
    override val message: String = "",
) : CompleteApiError<T>(), MessageError
```
But if we want more specific classes yours looks good! We should change it to _ApiResponse_ though, shouldn't we?

Reply:
Yup sounds good! I will do a follow-up PR with this change in API layer to minimize the changes in this PR

Reply:
Side note, <Any?> in these cases can be replaced using <Nothing> instead.